crates/libs/windows/src/core/abi.rs

```diff
@@ -1,48 +1,73 @@
 use super::*;
 
+/// Describes how a type can be transfered across FFI boundaries
+///
+/// # Safety
+///
+/// * The associated type `Abi` must be safe to transfer over FFI boundaries (e.g., it must have a stable layout)
+/// * `from_abi` must be implemented if there are some in-memory representations of `Abi` that are not valid representations of `Self`
 #[doc(hidden)]
 pub unsafe trait Abi: Sized {
+    /// The type used to represent `Self` across FFI boundaries
     type Abi;
 
+    /// Converts an abi value to `Self` or fails
+    ///
     /// # Safety
+    ///
+    /// This function is safe to call if `abi` can safely be trivial transmuted to `Self`
     unsafe fn from_abi(abi: Self::Abi) -> Result<Self> {
         Ok(core::mem::transmute_copy(&abi))
```

Comment:
It's always safe to convert from a valid ABI type via transmute. Keep in mind that this is only intended for the `windows` crate to use internally and scoped to Windows types that follow a certain ABI convention that makes this safe. 

While a `derive` macro would solve so many problems, the `derive` macros are insanely slow which is why I've scrubbed all but the essentials (`PartialEq` and `Eq` required for constant patterns) out of the `windows` crate. 

Of course, if a certain compiler dev wants to speed up the `derive` macro then I'm all for it. 😄

Reply:
> It's always safe to convert from a valid ABI type via transmute.

This isn't really true. Converting a null pointer to a `T: Interface` is incorrect (and thus why the implementation of `Abi for T where T: Interface` explicitly checks for null), so there's a good reason for there to be an explicit `from_abi` method. 

The reverse (i.e., going from `T` to `T::abi`) is currently always done via transmute. I'd personally be in favor of adding a more explicit `to_abi` method though. This would make the role of the `Abi` trait much clear. It would make verifying safety much more straightforward, and it would eliminate some calls to `std::mem::transmute` which I think is a goal we should always have. 

So, I agree that a `derive` macro probably isn't worth implementation complexity and resulting compile time slow down, but I do think it would be worth writing a *little* bit more code that makes all of this much more straightforward to understand.

Reply:
> This isn't really true.

Sorry if that was misleading. I was giving the short-hand C++ explanation. The Rust explanation is a little more nuanced only because we distinguish between nullable types and non-nullable types which we don't do in C++. You'll see this in `RuntimeType::DefaultType`. So in Rust, *it's always safe to convert from a valid ABI type via transmute to a corresponding "default type"*. I do need to merge/harmonize the `Abi` trait and the `RuntimeType` trait so let's avoid overhauling the `Abi` trait without keeping that in mind. Happy to consider pairing `to_abi` and `from_abi` - the confusion just comes from the split between the `Abi` and `RuntimeType` but I think it will be a lot simpler to understand once those are combined.

Reply:
Just for my own understanding, what is `RuntimeType` used for now? What do you envision things looking like after merging/harmonizing `Abi` and `RuntimeType`?

Reply:
`RuntimeType` was originally required beyond just `Abi` because WinRT includes generic COM interfaces where certain method signatures may have different types and semantics based on the generic type parameters and if those type parameters are nullable then need to know what the default type ends up being in order to build the projection correctly. It's all rather convoluted and more so in Rust because of `Option`/`NonNull`. Then throw ABI safety with nullable types and the need for a memory layout compatible borrow type it makes sense to harmonize this, which is what I've been meaning to do for some time. So let's hold off on any major `Abi` trait changes until I've got that sorted out and then I can get you guys to look it over.